app/controllers/pensions/overseas/incomeFromOverseasPension/PensionPaymentsController.scala

```diff
@@ -0,0 +1,124 @@
+  extends FrontendController(mcc) with I18nSupport with SessionHelper with FormUtils {
+
+  def amountForm(user: User): Form[(Option[BigDecimal], Option[BigDecimal])] = OptionalTupleAmountForm.amountForm(
+    emptyFieldKey1 = "overseasPension.pensionPayments.amountBeforeTax.noEntry",
+    wrongFormatKey1 = s"overseasPension.pensionPayments.amountBeforeTax.incorrectFormat.${if (user.isAgent) "agent" else "individual"}",
+    exceedsMaxAmountKey1 = "overseasPension.pensionPayments.amountBeforeTax.tooBig",
+    emptyFieldKey2 = "common.pensions.error.amount.noEntry",
+    wrongFormatKey2 = "overseasPension.pensionPayments.nonUkTaxPaid.incorrectFormat",
+    exceedsMaxAmountKey2 = "common.pensions.error.amount.overMaximum"
+  )
+
+  def show(taxYear: Int, index: Option[Int]): Action[AnyContent] = (authAction andThen taxYearAction(taxYear)).async { implicit request =>
+    pensionSessionService.getPensionSessionData(taxYear, request.user).map {
+      case Right(Some(data)) =>
+        validateIndex(index, data.pensions.incomeFromOverseasPensions.overseasIncomePensionSchemes) match {
+          case Some(i) => populateForm(data, taxYear, i)
+          case None => Redirect(PensionsSummaryController.show(taxYear)) // Todo should redirect to another page
+        }
+      case _ =>
+        //TODO: - Redirect to Overseas Pensions cya page??
+        Redirect(PensionsSummaryController.show(taxYear))
+    }
+  }
+
+
+  def submit(taxYear: Int, index: Option[Int]): Action[AnyContent] = authAction.async { implicit request =>
+
+    pensionSessionService.getPensionSessionData(taxYear, request.user).flatMap {
+      case Right(Some(data)) =>
+        validateIndex(index, data.pensions.incomeFromOverseasPensions.overseasIncomePensionSchemes) match {
+          case Some(i) => amountForm(request.user).bindFromRequest.fold(
+            formWithErrors => Future.successful(BadRequest(pensionPaymentsView(formWithErrors, taxYear, i))),
+            amounts =>
+              updatePensionScheme(data, amounts, taxYear, i))
+          case None => Future.successful(Redirect(PensionsSummaryController.show(taxYear)))
+        }
+      case _ =>
+        //TODO: redirect to the lifetime allowance CYA page?
+        Future.successful(Redirect(PensionsSummaryController.show(taxYear)))
+    }
+  }
+
+
+  private def validateIndex(index: Option[Int], pensionSchemesList: Seq[PensionScheme]): Option[Int] = {
+    index.flatMap(i =>
+      if (i >= 0 && i < pensionSchemesList.size) Some(i) else None
+    )
+  }
+
+  private def populateForm(data: PensionsUserData, taxYear: Int, index: Int)(implicit request: AuthorisationRequest[AnyContent]): Result = {
+    val amountBeforeTaxOpt = data.pensions.incomeFromOverseasPensions.overseasIncomePensionSchemes(index).pensionPaymentAmount
+    val nonUkTaxPaidOpt = data.pensions.incomeFromOverseasPensions.overseasIncomePensionSchemes(index).pensionPaymentTaxPaid
+
+    (amountBeforeTaxOpt, nonUkTaxPaidOpt) match {
```

Comment:
You could extract the common wrapper of `Ok(pensionPaymentsView)` as all this only generates the form.

Because: less duplication.

For instance:

```
 val form = (amountBeforeTaxOpt, nonUkTaxPaidOpt) match {
      case (Some(amountBeforeTax), None) => amountForm(request.user).fill(Some(amountBeforeTax), None)
      case (None, Some(nonUkTaxPaid)) => amountForm(request.user).fill((None, Some(nonUkTaxPaid)))
      case (Some(amountBeforeTax), Some(nonUkTaxPaid)) => amountForm(request.user).fill((Some(amountBeforeTax), Some(nonUkTaxPaid)))
      case (_, _) => amountForm(request.user)
    }
    Ok(pensionPaymentsView(form, taxYear, index))
```

Reply:
Fixed.

Reply:
Confirmed.